src/compiler/checker.ts

```diff
@@ -410,6 +410,7 @@ namespace ts {
                 const location = getParseTreeNode(locationIn);
                 return location ? getTypeOfSymbolAtLocation(symbol, location) : errorType;
             },
+            getNonMissingTypeOfSymbol,

```

Comment:
@sandersn Thanks for the feedback. I think `getTypeOfSymbol` will work. I wanted to use `getTypeOfSymbolAtLocation`, however for this case we don't have declarations., I checked `getTypeOfSymbolAtLocation` code and it uses `getNonMissingTypeOfSymbol`, so I just exported it. I'll push changes in two minutes.

Reply:
I've changed to `getTypeOfSymbol`

Reply:
@sandersn I've changed to `getTypeOfSymbol`

Reply:
Earlier, I didn't think about the way this PR would expose both `getTypeOfSymbol` and `getTypeOfSymbolAtLocation` in the API. This is confusing, but I guess it's all right since `getTypeOfSymbol` is marked internal. Still, what is the problem with `getTypeOfSymbolAtLocation`? Is it something that should be fixed so everything that uses it will get the improvement?

Reply:
getTypeOfSymbolAtLocation requires to use declaration as a second argument. I can use parent declaration. Or we can make second argument as optional, getNonMissingTypeOfSymbol doesn’t require it.

Reply:
`getTypeOfSymbolAtLocation` requires to use `node`
```ts
getTypeOfSymbolAtLocation(symbol: Symbol, node: Node): Type;
```

In the following case, properties do not have "declarations", so I see three ways to solve this problem
```ts
const names: { [K in D]: K } = {};
```

1. use `getTypeOfSymbolAtLocation` with `node` that somehow relate to property, for instance `parentDeclaration`, even it is not `prop` declaration. 

```ts
```
It is allowed because `getTypeOfSymbolAtLocation` doesn't completely depend on this `node`. In this case, we skip all conditions in getTypeOfSymbolAtLocation and get the result from `getNonMissingTypeOfSymbol`. However, I think this looks like a workaround instead of explicitly calling `getNonMissingTypeOfSymbol`.

```ts
function getTypeOfSymbolAtLocation(symbol: Symbol, location: Node) {
    symbol = symbol.exportSymbol || symbol;

    // If we have an identifier or a property access at the given location, if the location is
    // an dotted name expression, and if the location is not an assignment target, obtain the type
    // of the expression (which will reflect control flow analysis). If the expression indeed
    // resolved to the given symbol, return the narrowed type.
    if (location.kind === SyntaxKind.Identifier || location.kind === SyntaxKind.PrivateIdentifier) {
        if (isRightSideOfQualifiedNameOrPropertyAccess(location)) {
            location = location.parent;
        }
        if (isExpressionNode(location) && (!isAssignmentTarget(location) || isWriteAccess(location))) {
            const type = getTypeOfExpression(location as Expression);
            if (getExportSymbolOfValueSymbolIfExported(getNodeLinks(location).resolvedSymbol) === symbol) {
                return type;
            }
        }
    }
    if (isDeclarationName(location) && isSetAccessor(location.parent) && getAnnotatedAccessorTypeNode(location.parent)) {
        return resolveTypeOfAccessors(location.parent.symbol, /*writing*/ true) !;
    }
    // The location isn't a reference to the given symbol, meaning we're being asked
    // a hypothetical question of what type the symbol would have if there was a reference
    // to it at the given location. Since we have no control flow information for the
    // hypothetical reference (control flow information is created and attached by the
    // binder), we simply return the declared type of the symbol.
    return getNonMissingTypeOfSymbol(symbol);
}
```

Reply:
`getTypeOfSymbolAtLocation` requires to use `node`
```ts
getTypeOfSymbolAtLocation(symbol: Symbol, node: Node): Type;
```

In the following case, properties don't have "declarations", so I see three ways to solve this issue
```ts
const names: { [K in D]: K } = {};
```

1. use `getTypeOfSymbolAtLocation` with `node` that somehow relate to property, for instance `parentDeclaration`, even it is not `prop` declaration. 

```ts
checker.getTypeOfSymbolAtLocation(prop, prop.valueDeclaration ?? info.parentDeclaration);
```
It is allowed because `getTypeOfSymbolAtLocation` doesn't completely depend on `node`. In this case, we skip all conditions in `getTypeOfSymbolAtLocation` and get the result from `getNonMissingTypeOfSymbol`. However, I think this looks like a workaround instead of explicitly calling `getNonMissingTypeOfSymbol`.

2. export `getNonMissingTypeOfSymbol`/`getTypeOfSymbol` and use it without reliance on declaration
3. Not sure, maybe we can allow `node` to be `undefined` in `getTypeOfSymbolAtLocation`.

Reply:
Coming back to this, I think that (2), exporting `getTypeOfSymbol`, is an OK solution...but mostly because I'd like to use `getTypeOfSymbol` myself sometimes. =)

`getTypeOfSymbolAtLocation` should really support synthetic symbols that come from mapped properties, but it's a tricky design question how. Ideally the user would provide the original properties' declaration, but I'm not sure how they could obtain that.